torch/csrc/utils/tensor_new.cpp

```diff
@@ -224,34 +223,78 @@ Tensor internal_new_from_data(
     return var.to(device, inferred_scalar_type, /*non_blocking=*/false, /*copy=*/copy_variables);
   }
 
+  Tensor tensor;
 #ifdef USE_NUMPY
-  if (PyObject_HasAttrString(data, "__cuda_array_interface__")) {
-    TORCH_CHECK(!pin_memory, "Can't pin tensor constructed from __cuda_array_interface__");
-    auto tensor = tensor_from_cuda_array_interface(data);
-    const auto& inferred_scalar_type = type_inference ? tensor.scalar_type() : scalar_type;
-    auto device = device_opt.has_value() ? *device_opt : options.device();
-    pybind11::gil_scoped_release no_gil;
-    maybe_initialize_cuda(device);
-    return tensor.to(device, inferred_scalar_type, /*non_blocking=*/false, /*copy=*/copy_numpy);
-  }
+  if (is_numpy_available()) {
+    if (PyArray_Check(data)) {
+      TORCH_CHECK(!pin_memory, "Can't pin tensor constructed from numpy");
+      tensor = tensor_from_numpy(data, /*warn_if_not_writeable=*/!copy_numpy);
+    } else {
+      // CUDA/CPU Array Interface support.
+      //
+      // We are not using PyObject_HasAttrString because it suppresses
+      // errors that may be raised inside the __array_interface__
+      // property. Although, numpy suppresses errors, it is considered
+      // a bug, see numpy issue #11629.
+      PyObject* attr = PyObject_GetAttrString(data, "__cuda_array_interface__");
```

Comment:
What's the impact on performance from these additional checks? 

Should we consider limiting this functionality to torch.from_numpy()?

Reply:
> What's the impact on performance from these additional checks?

The suggested code is more efficient than the original code because `HasAttrString` is avoided.

> Should we consider limiting this functionality to torch.from_numpy()?

`torch.from_numpy()` inputs can only be NumPy ndarray objects but not objects that implement the Array Interface protocol. So, no (IIUC the question).